libraries/chain/db_block.cpp

```diff
@@ -103,6 +140,32 @@ std::vector<block_id_type> database::get_block_ids_on_fork(block_id_type head_of
   result.emplace_back(branches.first.back()->previous_id());
   return result;
 }
+    
+void database::check_tansaction_for_duplicated_operations(const signed_transaction& trx)
+{
+   const auto& proposal_index = get_index<proposal_object>();
+   std::set<fc::sha256> existed_operations_digests;
+   
+   proposal_index.inspect_all_objects( [&](const object& obj){
+      const proposal_object& proposal = static_cast<const proposal_object&>(obj);
+      for (auto& operation: proposal.proposed_transaction.operations)
+      {
+         existed_operations_digests.insert(fc::digest(operation));
+      }
+   });
+   
+   for (auto& pending_transaction: _pending_tx)
+   {
+      auto proposed_operations_digests = gather_proposed_operations_digests(pending_transaction);
+      existed_operations_digests.insert(proposed_operations_digests.begin(), proposed_operations_digests.end());
+   }
+    
+   auto proposed_operations_digests = gather_proposed_operations_digests(trx);
+   for (auto& digest: proposed_operations_digests)
+   {
+      FC_ASSERT(existed_operations_digests.count(digest) == 0, "Proposed operation is already pending for approval.");
```

Comment:
Note that this opens up a vector for DOS attacks - someone could create a proposal containing an operation he doesn't want to happen, and make sure it never gets approved.

Reply:
Right now, I am unsure how to best deal with this.An option could be to let witnesses use a separate api call `broadcast_transaction_dupcheck` that contains the limitations and leave the regular one unchanged.

Reply:
What is the purpose of this anyway?
Since it operates on the API level it can only be effective against accidental duplication. Witnesses are probably capable of circumventing this if they want to.

Reply:
Out initial motivation was to compare with pending transaction to reduce race conditions where multiple proposers are triggered at (more or less) the same to, to produce a proposal.
We only need one proposal and other witnesses are supposed to approve only. Hence a dupcheck against pending transaction. We also do a dupcheck against existing proposals which should help us as well to not end up with multiple prosals propsing the same thing

Reply:
My recommendation would be to handle that in the client.